### PR TITLE
Support fragmented packets

### PR DESCRIPTION
**Pull request by BonnyAD9**

### CODE REVIEW

File: input/parser.cpp [hunk 1]
```diff
@@ -299,7 +306,13 @@ uint16_t skip_ipv6_ext_hdrs(const u_char *data_ptr, uint16_t data_len, Packet *p
       } else if (next_hdr == IPPROTO_AH) {
          hdrs_len += (ext->ip6e_len << 2) - 2;
       } else if (next_hdr == IPPROTO_FRAGMENT) {
-         hdrs_len += 8;
+         // extract the fragmentation info
+         auto *frag = reinterpret_cast<const ip6_frag *>(data_ptr + hdrs_len);
+         pkt->frag_id = ntohl(frag->frag_id);
+         pkt->frag_off = ntohs(frag->frag_off) & IPV6_FRAGMENT_OFFSET;
```
[the remark this answers is not in the document]
Comment (BonnyAD9) [reply input/parser.cpp:312]: The field `ip6_frag::frag_off` is not only the fragment offset, the fragment offset is only the 13 MSb, the 3 LSb have different meaning. So to get the value in octets you would shift right by 3. And than to get the value in bytes you would shift left by 3. The two shifts cancel so you only need the mask.

This is also what the comment at `input/headers.hpp:54` tries to say.

File: storage/fragmentationCache/fragmentCache.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        || a.frag_id != b.frag_id)
+    {
+        return false;
+    }
+
+    if (a.ipv == IP::v4) {
+        return a.src_ip.v4 == b.src_ip.v4 && a.dst_ip.v4 == b.dst_ip.v4;
+    }
+    return memcmp(a.src_ip.v6, b.src_ip.v6, sizeof(a.src_ip.v6)) == 0
+        && memcmp(a.dst_ip.v6, b.dst_ip.v6, sizeof(a.src_ip.v6)) == 0;
```
[the remark this answers is not in the document]
Comment (BonnyAD9) [reply storage/fragmentationCache/fragmentCache.cpp:118]: Yes it can be done with a single memcmp, I did it this way because when the `ipv` is ipv4 you only need to compare the first four bytes of the `src_ip` and `dst_ip` because other bytes are 0.

File: input/parser.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Lukas955) [reply input/parser.cpp:312]: Ok, now it make sense. I suggest improving the comment to include this hidden logic.

File: storage/fragmentationCache/fragmentCache.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Lukas955) [reply storage/fragmentationCache/fragmentCache.cpp:118]: Yeah, I get it. However, from a CPU execution point of view, it would be much more optimal to use `memcmp()`:
- the compiler will be able to use fast vector units (i.e. SSE/AVX instructions)
- at the same time, most of the code branching will be removed, which in turn eliminates the impact of brach-predictors in the processor. This is because modern processors execute branches speculatively, and if there is a chance that the branching path of code execution changes frequently, this can have a significant impact on performance.

In other words, I would replace whole body of the function with a simple `memcmp()`.

File: input/parser.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Lukas955) [reply input/parser.cpp:312]: In your fix you have: "The fragment offset is in octets so to get the value in bytes you only need...", where the word "octets" is usually synonym for "bytes", so it is still confusing. You probably what to say: `The fragment offset is in 8-octet units so to get the value in bytes...`.

File: include/ipfixprobe/packet.hpp [hunk 3]
```diff
@@ -59,7 +59,10 @@ struct Packet : public Record {
    uint8_t     ip_flags;
    ipaddr_t    src_ip;
    ipaddr_t    dst_ip;
-   uint32_t    vlan_id;
+   uint16_t    vlan_id;
+   uint32_t    frag_id;
+   uint16_t    frag_off;
```
Comment (Lukas955) [inline include/ipfixprobe/packet.hpp:64]: For an IPv6 packet without fragmentation header, `frag_off` and `frag_id` values are undefined as you don't clear them in the constructor.

**Comment by SiskaPavel** [comment]
Moved to #180
